src/backend/drm/compositor/mod.rs

```diff
@@ -0,0 +1,3113 @@
+//! is copied over to the allocated gbm buffer for the cursor plane.
+//!
+//! Note: While the [`DrmCompositor`] also works on *legacy* drm the use of overlay and cursor planes is disabled in that case.
+//! Direct scan-out will only work with an atomic [`DrmSurface`].
+//!
+//! ## What makes a [`Element`] eligible for direct scan-out
+//!
+//! ### General
+//!
+//! First the element has to provide a [`UnderlyingStorage`] which can be exported as a drm framebuffer.
+//! Currently this is limited to wayland buffers, but may be extended in the future.
+//! This module provides a default exporter based on [`gbm`] which should fit most use-cases.
+//!
+//! If a certain combination of elements works can only be determined by asking the driver by submitting
+//! a atomic commit test. If that test fails the element is scheduled to be rendered on the primary plane.
+//!
+//! ### Overlay planes
+//!
+//! The element can only be directly scanned out if it's geometry does not overlap with an already assigned
+//! element on a plane higher in the stack.
+//!
+//! ### Underlay planes
+//!
+//! An underlay plane is only used if it does not overlap with an already assigned plane lower in the stack
+//! and the element is fully opaque.
+//!
+//! ### Primary plane
+//!
+//! For an element to be considered to be directly scanned out on the primary plane it has to be the last remaining
+//! visible element on the output and no other element has been assigned to the primary plane. If there are multiple
+//! element assigned to the primary plane the renderer will be used to composite the primary plane into a allocator
+//! provided buffer. Additionally the element has to be either fully opaque or the clear color has to match the CRTC
+//! background color and no overlap with an underlay is found.
+//!
+//! # How to use it
+//!
+//! ```no_run
+//! # use crate::backend::{
+//! #     allocator::gbm::{GbmAllocator, GbmDevice},
+//! #     drm::{DrmDevice, DrmDeviceFd},
+//! #     renderer::{
+//! #       element::surface::WaylandSurfaceRenderElement,
+//! #       gles2::{Gles2Renderbuffer, Gles2Renderer},
+//! #     },
+//! # };
+//! # use drm_fourcc::{DrmFormat, DrmFourcc, DrmModifier};
+//! # use std::{collections::HashSet, mem::MaybeUninit};
+//! #
+//! use smithay::{
+//!     backend::drm::{compositor::DrmCompositor, DrmSurface},
+//!     output::{Output, PhysicalProperties, Subpixel},
+//!     utils::Size,
+//! };
+//!
+//! // ...initialize the output, drm device, drm surface and allocator
+//! #
+//! # const CLEAR_COLOR: [f32; 4] = [0f32, 0f32, 0f32, 0f32];
+//! #
+//! let output = Output::new(
+//!     "e-DP".into(),
+//!     PhysicalProperties {
+//!         size: Size::from((800, 600)),
+//!         make: "N/A".into(),
+//!         model: "N/A".into(),
+//!         subpixel: Subpixel::Unknown,
+//!     },
+//!     None,
+//! );
+//!
+//! # let device: DrmDevice = unsafe { MaybeUninit::uninit().assume_init() };
```

Comment:
We should probably just use `todo!()` here, I'm a little scared by using MaybeUninit here. Plus the example is `no_run`, so it shouldn't cause any real change.

Reply:
Yeah, I have also used `unreachable!()` in those cases.

Reply:
Oh, I never thought about using just `todo!()` and always used some obscure ways, thanks!